// @web5/crypto/dist/cjs/crypto-primitives/pbkdf2.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pbkdf2 = void 0;
var crypto_1 = require("@noble/hashes/crypto");
var utils_js_1 = require("../utils.js");
/**
 * The `Pbkdf2` class provides a secure way to derive cryptographic keys from a password
 * using the PBKDF2 (Password-Based Key Derivation Function 2) algorithm. This class
 * supports both the Web Crypto API and Node.js Crypto module to offer flexibility in
 * different JavaScript environments.
 *
 * The PBKDF2 algorithm is widely used for generating keys from passwords, as it applies
 * a pseudorandom function to the input password along with a salt value and iterates the
 * process multiple times to increase the key's resistance to brute-force attacks.
 *
 * This class offers a single static method `deriveKey` to perform key derivation. It
 * automatically chooses between Web Crypto and Node.js Crypto based on the runtime
 * environment's support.
 *
 * Usage Examples:
 *
 * ```ts
 * const options = {
 *   hash: 'SHA-256', // The hash function to use ('SHA-256', 'SHA-384', 'SHA-512')
 *   password: new TextEncoder().encode('password'), // The password as a Uint8Array
 *   salt: new Uint8Array([...]), // The salt value
 *   iterations: 1000, // The number of iterations
 *   length: 256 // The length of the derived key in bits
 * };
 * const derivedKey = await Pbkdf2.deriveKey(options);
 * ```
 *
 * @remarks
 * This class relies on the availability of the Web Crypto API or Node.js Crypto module.
 * It falls back to Node.js Crypto if Web Crypto is not supported in the environment.
 */
var Pbkdf2 = /** @class */ (function () {
    function Pbkdf2() {
    }
    /**
     * Derives a cryptographic key from a password using the PBKDF2 algorithm.
     *
     * This method applies the PBKDF2 algorithm to the provided password along with
     * a salt value and iterates the process a specified number of times. It uses
     * a cryptographic hash function to enhance security and produce a key of the
     * desired length. The method is capable of utilizing either the Web Crypto API
     * or the Node.js Crypto module, depending on the environment's support.
     *
     * Example usage:
     *
     * ```ts
     * const options = {
     *   hash: 'SHA-256',
     *   password: new TextEncoder().encode('password'),
     *   salt: new Uint8Array([...]),
     *   iterations: 1000,
     *   length: 256
     * };
     * const derivedKey = await Pbkdf2.deriveKey(options);
     *
     * @param options - The options for key derivation.
     * @param options.hash - The hash function to use, such as 'SHA-256', 'SHA-384', or 'SHA-512'.
     * @param options.password - The password from which to derive the key, represented as a Uint8Array.
     * @param options.salt - The salt value to use in the derivation process, as a Uint8Array.
     * @param options.iterations - The number of iterations to apply in the PBKDF2 algorithm.
     * @param options.length - The desired length of the derived key in bits.
     *
     * @returns A Promise that resolves to the derived key as a Uint8Array.
     * ```
     */
    Pbkdf2.deriveKey = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if ((0, utils_js_1.isWebCryptoSupported)()) {
                    return [2 /*return*/, Pbkdf2.deriveKeyWithWebCrypto(options)];
                }
                else {
                    return [2 /*return*/, Pbkdf2.deriveKeyWithNodeCrypto(options)];
                }
                return [2 /*return*/];
            });
        });
    };
    Pbkdf2.deriveKeyWithNodeCrypto = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var password, salt, iterations, hashToNodeCryptoMap, hash, length, pbkdf2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        password = options.password, salt = options.salt, iterations = options.iterations;
                        hashToNodeCryptoMap = {
                            'SHA-256': 'sha256',
                            'SHA-384': 'sha384',
                            'SHA-512': 'sha512'
                        };
                        hash = hashToNodeCryptoMap[options.hash];
                        length = options.length / 8;
                        return [4 /*yield*/, import('node:crypto')];
                    case 1:
                        pbkdf2 = (_a.sent()).pbkdf2;
                        return [2 /*return*/, new Promise(function (resolve) {
                                pbkdf2(password, salt, iterations, length, hash, function (err, derivedKey) {
                                    if (!err) {
                                        resolve(new Uint8Array(derivedKey));
                                    }
                                });
                            })];
                }
            });
        });
    };
    Pbkdf2.deriveKeyWithWebCrypto = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var hash, password, salt, iterations, length, webCryptoKey, derivedKeyBuffer, derivedKey;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        hash = options.hash, password = options.password, salt = options.salt, iterations = options.iterations, length = options.length;
                        return [4 /*yield*/, crypto_1.crypto.subtle.importKey('raw', password, { name: 'PBKDF2' }, false, ['deriveBits'])];
                    case 1:
                        webCryptoKey = _a.sent();
                        return [4 /*yield*/, crypto_1.crypto.subtle.deriveBits({ name: 'PBKDF2', hash: hash, salt: salt, iterations: iterations }, webCryptoKey, length)];
                    case 2:
                        derivedKeyBuffer = _a.sent();
                        derivedKey = new Uint8Array(derivedKeyBuffer);
                        return [2 /*return*/, derivedKey];
                }
            });
        });
    };
    return Pbkdf2;
}());
exports.Pbkdf2 = Pbkdf2;
