// @web5/crypto/dist/cjs/crypto-algorithms/aes-ctr.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AesCtrAlgorithm = void 0;
var index_js_1 = require("../crypto-primitives/index.js");
var index_js_2 = require("../algorithms-api/index.js");
var AesCtrAlgorithm = /** @class */ (function (_super) {
    __extends(AesCtrAlgorithm, _super);
    function AesCtrAlgorithm() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.names = ['A128CTR', 'A192CTR', 'A256CTR'];
        return _this;
    }
    AesCtrAlgorithm.prototype.decrypt = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var algorithm, key, data, plaintext;
            return __generator(this, function (_a) {
                algorithm = options.algorithm, key = options.key, data = options.data;
                // Validate the input parameters.
                this.checkDecryptOptions(options);
                plaintext = index_js_1.AesCtr.decrypt({
                    counter: algorithm.counter,
                    data: data,
                    key: key,
                    length: algorithm.length
                });
                return [2 /*return*/, plaintext];
            });
        });
    };
    AesCtrAlgorithm.prototype.encrypt = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var algorithm, key, data, ciphertext;
            return __generator(this, function (_a) {
                algorithm = options.algorithm, key = options.key, data = options.data;
                // Validate the input parameters.
                this.checkEncryptOptions(options);
                ciphertext = index_js_1.AesCtr.encrypt({
                    counter: algorithm.counter,
                    data: data,
                    key: key,
                    length: algorithm.length
                });
                return [2 /*return*/, ciphertext];
            });
        });
    };
    AesCtrAlgorithm.prototype.generateKey = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var algorithm, keyOperations, algorithmNameToLength, secretKey;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        algorithm = options.algorithm, keyOperations = options.keyOperations;
                        // Validate the input parameters.
                        this.checkGenerateKeyOptions({ algorithm: algorithm, keyOperations: keyOperations });
                        algorithmNameToLength = {
                            A128CTR: 128,
                            A192CTR: 192,
                            A256CTR: 256
                        };
                        return [4 /*yield*/, index_js_1.AesCtr.generateKey({ length: algorithmNameToLength[algorithm.name] })];
                    case 1:
                        secretKey = _a.sent();
                        if (secretKey) {
                            secretKey.alg = algorithm.name;
                            if (keyOperations)
                                secretKey.key_ops = keyOperations;
                            return [2 /*return*/, secretKey];
                        }
                        throw new Error('Operation failed: generateKey');
                }
            });
        });
    };
    return AesCtrAlgorithm;
}(index_js_2.BaseAesCtrAlgorithm));
exports.AesCtrAlgorithm = AesCtrAlgorithm;
